graph/db/sql_store.go

```diff
@@ -43,3 +99,768 @@ func NewSQLStore(db BatchedSQLQueries, kvStore *KVStore) *SQLStore {
+		ExtraOpaqueData: make([]byte, 0),
+	}
+
+	if len(dbNode.Signature) == 0 {
+		return node, nil
+	}
+
+	node.HaveNodeAnnouncement = true
+	node.AuthSigBytes = dbNode.Signature
+	node.Alias = dbNode.Alias.String
+	node.LastUpdate = time.Unix(dbNode.LastUpdate.Int64, 0)
+
+	var err error
+	node.Color, err = DecodeHexColor(dbNode.Color.String)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode color: %w", err)
+	}
+
+	// Fetch the node's features.
+	node.Features, err = getNodeFeatures(ctx, db, dbNode.ID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"features: %w", dbNode.ID, err)
+	}
+
+	// Fetch the node's addresses.
+	_, node.Addresses, err = getNodeAddresses(ctx, db, pub[:])
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"addresses: %w", dbNode.ID, err)
+	}
+
+	// Fetch the node's extra signed fields.
+	extraTLVMap, err := getNodeExtraSignedFields(ctx, db, dbNode.ID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"extra signed fields: %w", dbNode.ID, err)
+	}
+
+	recs, err := lnwire.CustomRecords(extraTLVMap).Serialize()
+	if err != nil {
+		return nil, fmt.Errorf("unable to serialize extra signed "+
+			"fields: %w", err)
+	}
+
+	if recs != nil {
+		node.ExtraOpaqueData = recs
+	}
+
+	return node, nil
+}
+
+// getNodeFeatures fetches the feature bits and constructs the feature vector
+// for a node with the given DB ID.
+func getNodeFeatures(ctx context.Context, db SQLQueries,
+	nodeID int64) (*lnwire.FeatureVector, error) {
+
+	rows, err := db.GetNodeFeatures(ctx, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get node(%d) features: %w",
+			nodeID, err)
+	}
+
+	features := lnwire.EmptyFeatureVector()
+	for _, feature := range rows {
+		features.Set(lnwire.FeatureBit(feature.Bit))
+	}
+
+	return features, nil
+}
+
+// getNodeExtraSignedFields fetches the extra signed fields for a node with the
+// given DB ID.
+func getNodeExtraSignedFields(ctx context.Context, db SQLQueries,
+	nodeID int64) (map[uint64][]byte, error) {
+
+	fields, err := db.GetExtraNodeTypes(ctx, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get node(%d) extra "+
+			"signed fields: %w", nodeID, err)
+	}
+
+	extraFields := make(map[uint64][]byte)
+	for _, field := range fields {
+		extraFields[uint64(field.Type)] = field.Value
+	}
+
+	return extraFields, nil
+}
+
+// upsertNode upserts the node record into the database. If the node already
+// exists, then the node's information is updated. If the node doesn't exist,
+// then a new node is created. The node's features, addresses and extra TLV
+// types are also updated. The node's DB ID is returned.
+func upsertNode(ctx context.Context, db SQLQueries,
+	node *models.LightningNode) (int64, error) {
+
+	params := sqlc.UpsertNodeParams{
+		Version: int16(ProtocolV1),
+		PubKey:  node.PubKeyBytes[:],
+	}
+
+	if node.HaveNodeAnnouncement {
+		params.LastUpdate = sql.NullInt64{
```

Comment:
nit: use `sqldb.SQLInt64()`, `SQLStr()` and `SQLTime()`.

Reply:
Sounds good to me. I assume precision down to one second is enough here? Otherwise we'd need to use `UnixNano()` in which case we should make it very clear in the documentation that it's nanoseconds and not seconds.

Reply:
yeah it is advertised in seconds in the announcement 👍